File: Map.java
import java.util.ArrayList;
import java.util.List;

public class Map {
    public List<Object>[][] grid;
    public List<Predator> predators;
    public List<Prey> prey;
    private int[][] predatorCount;
    private int[][] preyCount;

    private int predatorBorn = 0;
    private int preyBorn = 0;
    private int predatorDead = 0;
    private int preyEaten = 0;
    private int grassEaten = 0;
    private int grassGrown = 0;
    private int preyDead = 0;

    public void incrementPredatorBorn(){
        predatorBorn++;
    }
    public void incrementPreyBorn(){
        preyBorn++;
    }
    public void incrementPredatorDead(){
        predatorDead++;
    }
    public void incrementPreyEaten(){
        preyEaten++;
    }

    public void incrementGrassEaten(){
        grassEaten++;
    }
    public void incrementGrassGrown(){
        grassGrown++;
    }
    public void incrementPreyDead(){
        preyDead++;
    }
    public int getPredatorBorn(){
        return predatorBorn;
    }
    public int getPreyBorn(){
        return preyBorn;
    }
    public int getPredatorDead(){
        return predatorDead;
    }
    public int getPreyEaten(){
        return preyEaten;
    }
  
    public int getGrassEaten(){
        return grassEaten;
    }
    public int getGrassGrown(){
        return grassGrown;
    }
    public int getPreyDead(){
        return preyDead;
    }

    @SuppressWarnings("unchecked")
    public Map(int width, int height){
        grid = new ArrayList[width][height];
        predators = new ArrayList<>();
        prey = new ArrayList<>();
        predatorCount = new int[width][height];
        preyCount = new int[width][height];
        //prey = new ArrayList<>();
        for (int i = 0; i < width; i++){
            for (int j = 0; j < height; j++){
                grid[i][j] = new ArrayList<>();
                predatorCount[i][j] = 0;
                preyCount[i][j] = 0;
            }
        }
    }
    
    public int getWidth(){
        return grid.length;
    }

    public int getHeight(){
        return grid[0].length;
    }

    public boolean inBounds(int x, int y){
        return x >= 0 && y >= 0 && x < getWidth() && y < getHeight();
    }

    public List<Object> getEntitiesAt(int x, int y){
        if (inBounds(x, y)){
            return grid[x][y];
        }
        return null;
    }

    public boolean addEntityAt(int x, int y, Object entity){
        if (inBounds(x, y)){
            grid[x][y].add(entity);
            if (entity instanceof Predator) {
                predators.add((Predator) entity);
                predatorCount[x][y]++;
            }
            if (entity instanceof Prey) {
                prey.add((Prey) entity);
                preyCount[x][y]++;
            }
            return true;
        }
        return false;
    }

    public boolean removeEntityAt(int x, int y, Object entity){
        if (inBounds(x, y)){
            grid[x][y].remove(entity);
            if (entity instanceof Predator) {
                predators.remove((Predator) entity);
                predatorCount[x][y]--;
            }
            if (entity instanceof Prey) {
                prey.remove((Prey) entity);
                preyCount[x][y]--;
            }
            return true;
        }
        return false;
    }

    public boolean moveEntity(int fromX, int fromY, int toX, int toY, Object entity){
        if (inBounds(toX, toY)){
            grid[fromX][fromY].remove(entity);
            if (entity instanceof Predator) {
                predatorCount[fromX][fromY]--;
                predatorCount[toX][toY]++;
            } else if (entity instanceof Prey) {
                preyCount[fromX][fromY]--;
                preyCount[toX][toY]++;
            }
            grid[toX][toY].add(entity);
            return true;
        }
        return false;
    }

    public List<Predator> getAllPredators(){
        return predators;
    }

    public List<Prey> getAllPreys(){
        return prey;
    }


    public boolean hasMultiplePredatorsAt(int x, int y){
        if (inBounds(x, y)){
            return predatorCount[x][y] > 1;
        }
        return false;
    }

    public boolean hasMultiplePreysAt(int x, int y){
        if (inBounds(x, y)){
            return preyCount[x][y] > 1;
        }
        return false;
    }
    
    public boolean addGrassAt(int x, int y){
        if (!inBounds(x, y)){
            return false;
        }
        for (Object object : grid[x][y]){
            if(object instanceof Grass){
                return false;
            }
        }
        grid[x][y].add(new Grass());
        return true;
        
    }

    public boolean removeGrassAt(int x, int y){
        if (!inBounds(x, y)){
            return false;
        }

        for (Object object : grid[x][y]){
            if(object instanceof Grass){
                grid[x][y].remove(object);
                return true;
            }
        }
        return false;
    }

    public Grass getGrassAt(int x, int y) {
        for (Object object : grid[x][y]){
            if(object instanceof Grass){
                return (Grass) object;
            }
        }
        return null;
    }

    public boolean checkForGrassAt(int x, int y){
        for (Object object : grid[x][y]){
            if(object instanceof Grass){
                return true;
            }
        }
        return false;
    }

    public class Grass {
        
        private int nextGrow = 0;

        public int getNextGrow(){
            return nextGrow;
        }

        public void resetNextGrow(){
            nextGrow = 0;
        }

        private int nutrition = 0;

        public Grass(){
            nutrition = 2;
        }

        public int getNutrition(){
            return nutrition;
        }

        public void turn(){
            nextGrow++;
            if (nextGrow % 25 == 0){
            nutrition++;
            nextGrow = 0;
            Map.this.incrementGrassGrown();
            }
        }
        public void consume(){
            if (nutrition > 0){
                nutrition--;
            }
        }
    }

}

File: Map.js
class Map {
    constructor(width, height, Grass) {
        this.grid = Array(width).fill().map(() => Array(height).fill().map(() => []));
        this.predators = [];
        this.prey = [];
        this.predatorCount = Array(width).fill().map(() => Array(height).fill(0));
        this.preyCount = Array(width).fill().map(() => Array(height).fill(0));
        
        this.predatorBorn = 0;
        this.preyBorn = 0;
        this.predatorDead = 0;
        this.preyEaten = 0;
        this.grassEaten = 0;
        this.grassGrown = 0;
        this.preyDead = 0;

        this.Grass = Grass;
        
    }
      
    incrementPredatorBorn() {
        this.predatorBorn++;
    }
    
    incrementPreyBorn() {
      this.preyBorn++;
    }
    
    incrementPredatorDead() {
      this.predatorDead++;
    }
    
    incrementPreyEaten() {
      this.preyEaten++;
    }
    
    incrementGrassEaten() {
      this.grassEaten++;
    }
    
    incrementGrassGrown() {
      this.grassGrown++;
    }
    
    incrementPreyDead() {
      this.preyDead++;
    }

    getPredatorBorn(){
        return this.predatorBorn;
    }
    getPreyBorn(){
        return this.preyBorn;
    }
    getPredatorDead(){
        return this.predatorDead;
    }
    getPreyEaten(){
        return this.preyEaten;
    }
    getGrassEaten(){
        return this.grassEaten;
    }
    getGrassGrown(){
        return this.grassGrown;
    }
    getPreyDead(){
        return this.preyDead;
    }
    
    getWidth() {
        return this.grid.length;
    }
    
    getHeight() {
      return this.grid[0].length;
    }
    
    inBounds(x, y) {
      return x >= 0 && y >= 0 && x < this.getWidth() && y < this.getHeight();
    }
    
    getEntitiesAt(x, y) {
      if (this.inBounds(x, y)) {
        return this.grid[x][y];
      }
      return null;
    }
    // Add an entity to the map at the specified coordinates
    addEntityAt(x, y, entity) {
      if (this.inBounds(x, y)) {
        this.grid[x][y].push(entity);
        if (entity.constructor.name === 'Predator') {
          this.predators.push(entity);
          this.predatorCount[x][y]++;
        }
        if (entity.constructor.name === 'Prey') {
          this.prey.push(entity);
          this.preyCount[x][y]++;
        }
        return true;
      }
      return false;
    }
    // Remove an entity from the map at the specified coordinates
    removeEntityAt(x, y, entity) {
        if (this.inBounds(x, y)) {
          const index = this.grid[x][y].indexOf(entity);
          if (index !== -1) {
            this.grid[x][y].splice(index, 1);
            if (entity.constructor.name === 'Predator') {
              const predatorIndex = this.predators.indexOf(entity);
              if (predatorIndex !== -1) {
                this.predators.splice(predatorIndex, 1);
              }
              this.predatorCount[x][y]--;
            }
            if (entity.constructor.name === 'Prey') {
              const preyIndex = this.prey.indexOf(entity);
              if (preyIndex !== -1) {
                this.prey.splice(preyIndex, 1);
              }
              this.preyCount[x][y]--;
            }
            return true;
          }
        }
        return false;
    }
    // Move an entity from one position to another
    moveEntity(fromX, fromY, toX, toY, entity) {
      if (this.inBounds(toX, toY)) {
        const index = this.grid[fromX][fromY].indexOf(entity);
        if (index !== -1) {
          this.grid[fromX][fromY].splice(index, 1);
          this.grid[toX][toY].push(entity);
          
          if (entity.constructor.name === 'Predator') {
            this.predatorCount[fromX][fromY]--;
            this.predatorCount[toX][toY]++;
          } else if (entity.constructor.name === 'Prey') {
            this.preyCount[fromX][fromY]--;
            this.preyCount[toX][toY]++;
          }
          return true;
        }
      }
      return false;
    }

    getAllPredators() {
        return this.predators;
    }
      
    getAllPreys() {
        return this.prey;
    }

    // Get the number of predators at a specific position
    hasMultiplePredatorsAt(x, y) {
        if (this.inBounds(x, y)) {
          return this.predatorCount[x][y] > 1;
        }
        return false;
    }
    // Get the number of preys at a specific position
    hasMultiplePreysAt(x, y) {
        if (this.inBounds(x, y)) {
          return this.preyCount[x][y] > 1;
        }
        return false;
    }
    // Add grass at a specific position
    addGrassAt(x, y) {
        if (!this.inBounds(x, y)) {
          return false;
        }
        
        for (const object of this.grid[x][y]) {
          if (object.constructor.name === 'Grass') {
            return false;
          }
        }
        
        this.grid[x][y].push(new this.Grass(this));
        return true;
    }
    // Remove grass at a specific position
    removeGrassAt(x, y) {
      if (!this.inBounds(x, y)) {
        return false;
      }
      
      for (let i = 0; i < this.grid[x][y].length; i++) {
        const object = this.grid[x][y][i];
        if (object.constructor.name === 'Grass') {
          this.grid[x][y].splice(i, 1);
          return true;
        }
      }
      
      return false;
    }
    
    getGrassAt(x, y) {
      for (const object of this.grid[x][y]) {
        if (object.constructor.name === 'Grass') {
          return object;
        }
      }
      return null;
    }
    
    checkForGrassAt(x, y) {
      for (const object of this.grid[x][y]) {
        if (object.constructor.name === 'Grass') {
          return true;
        }
      }
      return false;
    }
}
    
export default Map;